Error on a page rendered via controller mapping

https://developer.enonic.com/search?q=webapp

If a page is not mapped to an underlying content item but rendered via controller mapping then SEO Metafields will give an error:
````
Caused by: org.openjdk.nashorn.internal.runtime.ECMAException: TypeError: Cannot get property "_path" of null
	at org.openjdk.nashorn.internal.runtime.ECMAErrors.error(ECMAErrors.java:57)
	at org.openjdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:212)
	at org.openjdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:184)
	at org.openjdk.nashorn.internal.runtime.ECMAErrors.typeError(ECMAErrors.java:171)
	at org.openjdk.nashorn.internal.runtime.linker.NashornBottomLinker.linkNull(NashornBottomLinker.java:258)
	at org.openjdk.nashorn.internal.runtime.linker.NashornBottomLinker.getGuardedInvocation(NashornBottomLinker.java:69)
	at jdk.dynalink/jdk.dynalink.linker.support.CompositeGuardingDynamicLinker.getGuardedInvocation(CompositeGuardingDynamicLinker.java:109)
	at jdk.dynalink/jdk.dynalink.LinkerServicesImpl.lambda$getGuardedInvocation$0(LinkerServicesImpl.java:137)
	at jdk.dynalink/jdk.dynalink.LinkerServicesImpl.getWithLookupInternal(LinkerServicesImpl.java:168)
	at jdk.dynalink/jdk.dynalink.LinkerServicesImpl.getGuardedInvocation(LinkerServicesImpl.java:135)
	at jdk.dynalink/jdk.dynalink.DynamicLinker.relink(DynamicLinker.java:241)
	at org.openjdk.nashorn.internal.scripts.Script$Recompilation$1043$901AA$add_metadata.L:1#responseProcessor(com.enonic.app.metafields:/site/processors/add-metadata.js:47)
	at org.openjdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:650)
	at org.openjdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:513)
	at org.openjdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:520)
	at org.openjdk.nashorn.api.scripting.ScriptObjectMirror.call(ScriptObjectMirror.java:111)
	at com.enonic.xp.script.impl.value.FunctionScriptValue.call(FunctionScriptValue.java:38)
```

--- a/jest.config.ts
+++ b/jest.config.ts
@@ -13,6 +13,7 @@ export default {
 	moduleNameMapper: {
 		'/guillotine/(.*)': '<rootDir>/src/main/resources/guillotine/$1',
 		'/lib/app-metafields/(.*)': '<rootDir>/src/main/resources/lib/app-metafields/$1',
+		'/site/(.*)': '<rootDir>/src/main/resources/site/$1',
 	},
 	preset: 'ts-jest/presets/js-with-babel-legacy',
 	testEnvironment: 'node',

--- a/package.json
+++ b/package.json
@@ -8,6 +8,7 @@
     "@enonic-types/lib-node": "^7.14.1",
     "@enonic-types/lib-portal": "^7.14.1",
     "@enonic/js-utils": "^1.7.1",
+    "@enonic/mock-xp": "^1.0.0",
     "@swc/core": "^1.4.17",
     "glob": "^10.3.12",
     "jest": "^29.7.0",

--- a/src/main/resources/lib/app-metafields/constants.ts
+++ b/src/main/resources/lib/app-metafields/constants.ts
@@ -3,4 +3,5 @@ import {getDashedAppName} from '/lib/app-metafields/xp/getDashedAppName';
 
 export const APP_NAME_PATH = getDashedAppName(); // Outside Guillotine Extension Context
 export const DEBUG = false;
+export const TRACE = false;
 export const MIXIN_PATH = 'meta-data';

--- a/src/main/resources/lib/app-metafields/types/Request.d.ts
+++ b/src/main/resources/lib/app-metafields/types/Request.d.ts
@@ -8,7 +8,8 @@ export declare type Request<
 		pathParams?: Record<string, string>
 		rawPath?: string
 		remoteAddress?: string
-		// webSocket?: unknown
+		repositoryId?: string
+		webSocket?: boolean
 	}
 > = {
 	branch: 'draft'|'master'

--- a/src/main/resources/site/processors/add-metadata.ts
+++ b/src/main/resources/site/processors/add-metadata.ts
@@ -8,7 +8,7 @@ import {
 	getSite as libPortalGetSite,
 	getSiteConfig as libPortalGetSiteConfig,
 } from '/lib/xp/portal';
-import {DEBUG} from '/lib/app-metafields/constants';
+import {DEBUG, TRACE} from '/lib/app-metafields/constants';
 import {getMergedConfig} from '/lib/app-metafields/xp/getMergedConfig';
 import {getFixedHtmlAttrsAsString} from '/lib/app-metafields/processor/getFixedHtmlAttrsAsString';
 import {getMetaData} from '/lib/app-metafields/processor/getMetaData'
@@ -20,8 +20,14 @@ const XML_MEDIA_TYPES = ['application/xhtml+xml', 'application/xml', 'text/xml']
 
 
 export const responseProcessor = (req: Request, res: Response) => {
+	DEBUG && log.debug('add-metadata response processor Request:%s Response:%s', toStr(req), toStr(res));
+
 	const content = getCurrentContent();
+	DEBUG && log.debug('add-metadata response processor Content:%s', toStr(content));
+
 	const site = libPortalGetSite();
+	DEBUG && log.debug('add-metadata response processor Site:%s', toStr(content));
+
 	const siteConfig = libPortalGetSiteConfig();
 	DEBUG && log.debug('add-metadata response processor siteConfig:%s', toStr(siteConfig));
 
@@ -29,6 +35,8 @@ export const responseProcessor = (req: Request, res: Response) => {
 
 	let titleAdded = false;
 	const isResponseContentTypeHtml = res.contentType.indexOf(HTML_MEDIA_TYPE) > -1;
+	TRACE && log.debug('add-metadata response processor isResponseContentTypeHtml:%s', isResponseContentTypeHtml);
+
 	const isResponseContentTypeXml = XML_MEDIA_TYPES.some(xmlMediaType => res.contentType.indexOf(xmlMediaType) > -1);
 
 	if ( isResponseContentTypeHtml && res.body && typeof res.body === "string" ) {
@@ -43,7 +51,8 @@ export const responseProcessor = (req: Request, res: Response) => {
 				mergedConfig,
 				content,
 				site,
-			}) || "";
+			}) || '';
+			TRACE && log.debug('add-metadata response processor titleHtml:%s', titleHtml);
 			res.body = res.body.replace(/(<title>)(.*?)(<\/title>)/i, titleHtml);
 			titleAdded = true;
 		}
@@ -52,7 +61,7 @@ export const responseProcessor = (req: Request, res: Response) => {
 		if (htmlIndex >= 0 && endHtmlIndex >= 0) {
 			const fixedHtmlTagInnerContent = getFixedHtmlAttrsAsString({
 				htmlTag: res.body,
-				isFrontpage: site._path === content._path,
+				isFrontpage: site && content && site._path === content._path,
 			});
 			res.body = res.body.substr(0, htmlIndex + 5) + " " + fixedHtmlTagInnerContent + res.body.substr(endHtmlIndex);
 		}

--- a/test/site/processors/add-metadata.test.ts
+++ b/test/site/processors/add-metadata.test.ts
@@ -1,0 +1,139 @@
+import type { Log } from '@enonic/mock-xp';
+import type {
+	getContent as libPortalGetContent,
+	getSite as libPortalGetSite,
+	getSiteConfig as libPortalGetSiteConfig
+} from '@enonic-types/lib-portal';
+import type {
+	Request,
+	Response
+} from '/lib/app-metafields/types';
+
+
+import {
+	// LibContent,
+	// LibContext,
+	// LibNode,
+	// LibPortal,
+	// App,
+	// Request as MockRequest,
+	Server
+} from '@enonic/mock-xp';
+import {
+	describe,
+	expect,
+	jest,
+	test as it,
+} from '@jest/globals';
+
+
+const server = new Server({
+	loglevel: 'info'
+});
+
+// Avoid type errors below.
+// eslint-disable-next-line @typescript-eslint/no-namespace
+declare module globalThis {
+	let log: Log
+}
+
+globalThis.log = server.log as Log;
+
+// const app = new App({ key: 'com.enonic.app.metafields' });
+// const libContent = new LibContent({ server });
+// const libContext = new LibContext({ server });
+// const libNode = new LibNode({ server });
+// const libPortal = new LibPortal({ app, server });
+
+jest.mock('/lib/xp/content', () => {
+    return {
+    }
+}, { virtual: true });
+
+jest.mock('/lib/xp/context', () => {
+    return {
+    }
+}, { virtual: true });
+
+jest.mock('/lib/xp/node', () => {
+    return {
+    }
+}, { virtual: true });
+
+jest.mock('/lib/xp/portal', () => {
+	return {
+		// assetUrl: jest.fn<typeof assetUrlType>((params) => libPortal.assetUrl(params)),
+		// getContent: jest.fn<typeof libPortalGetContent>(() => libPortal.getContent()),
+		getContent: jest.fn<typeof libPortalGetContent>(() => null),
+		// getSite: jest.fn<typeof libPortalGetSite>(() => libPortal.getSite()),
+		getSite: jest.fn<typeof libPortalGetSite>(() => null),
+		getSiteConfig: jest.fn<typeof libPortalGetSiteConfig<{}>>(() => ({
+			"canonical": true,
+			"blockRobots": false,
+			"twitterUsername": "@enonicHQ",
+			"seoImage": "f7efd6e4-9eed-45d2-83b6-d088a7bdc685",
+			"titleBehaviour": true,
+			"titleSeparator": "-",
+			"titleFrontpageBehaviour": true,
+			"disableAppConfig": false,
+			"seoImageIsPrescaled": false,
+			"frontpageImage": "f7efd6e4-9eed-45d2-83b6-d088a7bdc685",
+			"frontpageImageIsPrescaled": false,
+			"removeTwitterImage": false,
+			"removeOpenGraphUrl": false,
+			"removeOpenGraphImage": false,
+			"fullPath": false,
+			"headless": false
+		})),
+		// imageUrl: jest.fn<typeof imageUrlType>((params) => libPortal.imageUrl(params)),
+	}
+}, { virtual: true });
+
+jest.mock('/lib/thymeleaf', () => {
+	render: jest.fn((view, model) => model)
+}, { virtual: true });
+
+describe('add-metadata', () => {
+	it('should not throw when there is no content', () => {
+		import('/site/processors/add-metadata').then(({responseProcessor}) => {
+			const request: Request = {
+				"branch": "draft",
+				"method": "GET",
+				"scheme": "http",
+				"host": "localhost",
+				"port": 8080,
+				"path": "/admin/site/inline/dev-portal/draft/developer/search",
+				// "rawPath": "/admin/site/inline/dev-portal/draft/developer/search",
+				"url": "http://localhost:8080/admin/site/inline/dev-portal/draft/developer/search?q=test",
+				// "remoteAddress": "127.0.0.1",
+				"mode": "inline",
+				// "webSocket": false,
+				// "repositoryId": "com.enonic.cms.dev-portal",
+				// "contextPath": "/admin/site/inline/dev-portal/draft/developer",
+				// "params": {
+				// 	"q": "test"
+				// },
+			}
+			// libPortal.request = new MockRequest(request);
+			const response: Response = {
+				"status": 200,
+				"contentType": "text/html; charset=utf-8",
+				"postProcess": true,
+				"headers": {},
+				"pageContributions": {
+					"headEnd": [
+						"<script async src=\"/admin/site/inline/dev-portal/draft/_/asset/com.enonic.site.developer:1714396560884/js/main.min.js\"></script>"
+					]
+				},
+				"applyFilters": true,
+				"body": `<!DOCTYPE html><html lang="en"><head><title>Whatnot</title></head></html>`
+			};
+			const expected = {
+				...response,
+				body: `<!DOCTYPE html><html lang="en" prefix="og: http://ogp.me/ns# article: http://ogp.me/ns/article#"><head></head></html>`,
+			};
+			const actual = responseProcessor(request, response);
+			expect(actual).toStrictEqual(expected);
+		});
+	});
+});

